File: Rotting Oranges/Rotting Oranges.py
from typing import List
from collections import deque

class solution():
    def orangesRotting(self, grid:List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        queue = deque()
        fresh_oranges = 0
        rotten_count = 0
        minutes =0

        #Find initial rotten oranges and count fresh oranges
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 2:
                    queue.append((i,j))
                    rotten_count +=1
                elif grid[i][j] == 1:
                    fresh_oranges +=1
        
        # BFS algorithm
        while queue:
            size = len(queue)

            # process all oranges at current time
            for i in range(size):
                x, y = queue.popleft()

                #now explore all 4 directions of the cell (up, down, left, right)
                for dx, dy in [(0,1),(0,-1),(-1,0),(1,0)]:
                    new_x, new_y = x +dx, y +dy

                    # check if adjacent cell is a fresh orange
                    if 0 <= new_x < rows and 0 <= new_y < cols and grid[new_x, new_y]==1:
                        #if above condition satisfies its a fresh fruit make it as rotten
                        grid[new_x, new_y]=2
                        rotten_count += 1
                        fresh_oranges -= 1
                        queue.append((new_x, new_y))
            
            if rotten_count > 0:
                minutes +=1
        # Step 5: Check if there are any fresh oranges remaining        
        if fresh_oranges>0:
            return -1
        
        return minutes-1

File: Rotting Oranges/Rotting Oranges.cs
public class Solution {
    public int OrangesRotting(int[][] grid) {
        // Define the directions for adjacent cells (up, down, left, right)
        int[][] directions = new int[][] {
            new int[] {-1, 0},
            new int[] {1, 0},
            new int[] {0, -1},
            new int[] {0, 1}
        };

        int rows = grid.Length;
        int cols = grid[0].Length;
        int freshCount = 0;
        int minutes = 0;

        // Step 1: Initialize the queue and count the fresh oranges
        Queue<int[]> queue = new Queue<int[]>();
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 2) {  // Rotten orange
                    queue.Enqueue(new int[] {i, j});
                } else if (grid[i][j] == 1) {  // Fresh orange
                    freshCount++;
                }
            }
        }

        // Step 3: Perform BFS
        while (freshCount > 0 && queue.Count > 0) {
            int size = queue.Count;
            for (int i = 0; i < size; i++) {
                int[] orange = queue.Dequeue();
                int row = orange[0];
                int col = orange[1];
                foreach (int[] direction in directions) {
                    int newRow = row + direction[0];
                    int newCol = col + direction[1];
                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols &&
                        grid[newRow][newCol] == 1) {  // Fresh orange
                        grid[newRow][newCol] = 2;  // Mark as rotten
                        queue.Enqueue(new int[] {newRow, newCol});
                        freshCount--;
                    }
                }
            }
            if (queue.Count > 0) {
                minutes++;
            }
        }

        // Step 4: Check for remaining fresh oranges
        if (freshCount > 0) {
            return -1;
        }

        // Step 5: Return the minimum number of minutes
        return minutes;
    }
}
